lib/ioutil/log_reader.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ioutil
+
+import (
+	"bufio"
+	"io"
+
+	"github.com/mailru/easyjson/jlexer"
+)
+
+// LogReader unwraps entires into LogEntry structs.
+type LogReader struct {
+	io.ReadCloser
+	s *bufio.Scanner
+
+	// if the buffer we want to read into is too short for our log entry,
+	// we will need to return io.ErrShortBuffer and try again on the next call,
+	// necessitating this pointer to the previous LogEntry we attempted to use.
+	prev *LogEntry
+}
+
+// NewLogReader wraps an io.ReadCloser in a LogReader.
+func NewLogReader(r io.ReadCloser) *LogReader {
+	return &LogReader{
+		ReadCloser: r,
+		s:          bufio.NewScanner(r),
```

Comment:
The scanner is stored so that the position of the scanner can be preserved in the event that our buffer is too small to `copy()` everything and we need to return `io.ErrShortBuffer`. It's either that (or something like it), or we do some bookkeeping to hold onto remaining data to write into the next buffer provided in the next call. This seemed more straightforward.